ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryReplicaSyncUpManager.java

```diff
@@ -0,0 +1,213 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.clustermap;
+
+import com.github.ambry.config.ClusterMapConfig;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of {@link ReplicaSyncUpManager} that helps track replica catchup state.
+ */
+public class AmbryReplicaSyncUpManager implements ReplicaSyncUpManager {
+  private final ConcurrentHashMap<String, CountDownLatch> partitionToBootstrapLatch = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, Boolean> partitionToBootstrapSuccess = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<ReplicaId, LocalReplicaLagInfos> replicaToLagInfos = new ConcurrentHashMap<>();
+  private final ClusterMapConfig clusterMapConfig;
+
+  private static final Logger logger = LoggerFactory.getLogger(AmbryReplicaSyncUpManager.class);
+
+  public AmbryReplicaSyncUpManager(ClusterMapConfig clusterMapConfig) {
+    this.clusterMapConfig = clusterMapConfig;
+  }
+
+  @Override
+  public void initiateBootstrap(ReplicaId replicaId) {
+    partitionToBootstrapLatch.put(replicaId.getPartitionId().toPathString(), new CountDownLatch(1));
+    partitionToBootstrapSuccess.put(replicaId.getPartitionId().toPathString(), false);
+    replicaToLagInfos.put(replicaId,
+        new LocalReplicaLagInfos(replicaId, clusterMapConfig.clustermapReplicaCatchupAcceptableLagBytes));
+  }
+
+  @Override
+  public void waitBootstrapCompleted(String partitionName) throws InterruptedException {
+    CountDownLatch latch = partitionToBootstrapLatch.get(partitionName);
+    if (latch == null) {
+      logger.info("Skipping bootstrap for existing partition {}", partitionName);
+    } else {
+      logger.info("Waiting for new partition {} to complete bootstrap", partitionName);
+      latch.await();
+      partitionToBootstrapLatch.remove(partitionName);
+      if (!partitionToBootstrapSuccess.remove(partitionName)) {
+        throw new IllegalStateException("Partition " + partitionName + " failed on bootstrap.");
+      }
+      logger.info("Bootstrap is complete on partition {}", partitionName);
+    }
+  }
+
+  @Override
+  public boolean updateLagBetweenReplicas(ReplicaId source, ReplicaId target, long lagInBytes) {
+    boolean updated = false;
+    if (replicaToLagInfos.containsKey(source)) {
+      replicaToLagInfos.get(source).updateLagInfo(target, lagInBytes);
+      logger.debug(replicaToLagInfos.get(source).toString());
```

Comment:
You need to suround this inside a "if (Logger.isDebugEnabled())" check, otherwise the expensive toString() method will get called before the logger level is checked.

Reply:
fixed, thanks for the suggestion.